### ISSUE TITLE
XSS Problem with getHighlightedText in InputSuggestions

### ISSUE THREAD

**Issue by codefactor** [SAP/ui5-webcomponents#4245]
### **Bug Description**
There is a very serious XSS bug in the getHighlightedText functionality because the "sanitizeText" function is only replacing the first less than sign. It is not fully sanitizing the text.

https://github.com/SAP/ui5-webcomponents/blob/master/packages/main/src/features/InputSuggestions.js#L480

### **Expected Behavior**
Text for input and suggestion text are both considered "plain text" and not markup. Both texts should be sanitized, including not just less than signs, but other HTML special characters as well. You should even be able to highlight special characters like less than sign as well. But the logic today only sanitizes the text but not the search text so it doesn't find matches. There are other issues that might happen once you fix that issue as well so be very careful to add lots of unit tests with different possibilities.

### **Steps to Reproduce**
1. Have suggestions with text that contains markup with at least 1 dummy less than sign in beginning with XSS img tag
2. Get the XSS issue right away

### **Isolated Example**
https://codesandbox.io/s/ui5-webcomponents-forked-mymbd?file=/index.html

### **Context**
- UI5 Web Components version: latest
- OS/Platform: any
- Browser: any
- Affected component: InputSuggestions

### **Log Output / Stack Trace / Screenshots**
```
{...}
```
### **Priority**
- [ ] Low
- [x] Medium
- [ ] High
- [ ] Very High

The priority indicates the severity of the issue. To set the appropriate priority consider the following criteria:
- Breaks entire application or system - High or Very High
- Accessibility issue - Medium or High
- Functional issue - Medium or High
- Visual issue - Low or Medium

**Note:** The priority might be re-evaluated by the issue processor.

### **Stakeholder Info** (if applicable)

- Organization: {...}
- Bussiness impact: {...}

**Comment by codefactor** [comment]
Here is a little test script with at least what I am expecting:
```javascript
(function () {
    function getHighlightedText(suggestion, input) {
        let text = suggestion.text || suggestion.textContent;
        text = sanitizeText(text);

        return hightlightInput(text, input);
    }

    function hightlightInput(text, input) {
        if (!text) {
            return text;
        }

        const inputEscaped = input.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
        const regEx = new RegExp(inputEscaped, "ig");
        return text.replace(regEx, match => `<b>${match}</b>`);
    }

    function sanitizeText(text) {
        return text && text.replace("<", "&lt");
    }

    function test(text, input, expect) {
        const result = getHighlightedText({text}, input);
        if (result !== expect) {
            console.log(`failure: "${result}" was expecting "${expect}" for highlighting "${input}" on "${text}"`);
        } else {
            console.log(`success: "${result}" has highlighted "${input}" on "${text}" properly.`);
        }
    }

    test(`Scott`, `c`, `S<b>c</b>ott`);
    test(`<<img src="xss" onerror="alert('xss')">`, `<`, `<b>&lt;</b><b>&lt;</b>img src=&quot;xss&quot; onerror=&quot;alert(&apos;xss&apos;)&quot;>`);
})();
```

**Comment by codefactor** [comment]
The `text` attribute of the `<ui5-suggestion-item>` should contractually be considered raw text, and NOT markup. In other words, if the user has entered a text that contains characters like `<` or `&` these should be interpreted literally, rather than as HTML entities.

To summarize, there are several issues with this code:
1. The sanitize function only replaces the first occurrence of `<` symbol with `&lt;` instead of all of them
2. The sanitize function is ran on the input text prior to doing the find and replace, so it will break html entities in half by accident
3. The find/replace is done with the raw input text, so finds and replaces on top of markup so it won't highlight special characters if the input does contain something.

**Comment by codefactor** [comment]
Here is one implementation that I threw together:
```javascript
(function () {
    /** escape string to be used as regex as plain text. */
    const escapeRegex = (s) => s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");

    /** replace all occurrances of one find text with a replace text. */
    const replaceAll = (text, find, replace, flags) => {
        return text.replace(new RegExp(escapeRegex(find), `${flags||""}g`), replace);
    };

    /** make a token gauranteed not to be found in the input text. */
    const makeToken = (i, text) => {
        while (text.indexOf(i) >= 0) {
            i = `[${i}]`;
        }
        return i;
    };

    function getHighlightedText(suggestion, input) {
        let text = suggestion.text || suggestion.textContent;
        return hightlightInputMarkup(text, input);
    }

    function hightlightInputMarkup(text, input) {
        if (!text) {
            return text;
        }
        const start = makeToken('[b]', text);
        const end = makeToken('[/b]', text);
        const rawResult = replaceAll(text, input, match => `${start}${match}${end}`, 'i');
        return replaceAll(replaceAll(sanitizeText(rawResult), start, '<b>'), end, '</b>');
    }

    const replacements = `&=&amp,>=&gt;,<=&lt;,"=&quot;,'=&#39;`.split(",")
            .map((p) => p.split('='))
            .map(([f,r]) => [new RegExp(escapeRegex(f),'g'), r]);

    function sanitizeText(text) {
        let result = text;
        replacements.forEach(([regex, replacement]) => {
            result = result.replace(regex, replacement);
        });
        return result;
    }

    function test(text, input, expect) {
        const result = getHighlightedText({text}, input);
        if (result !== expect) {
            console.log(`failure:\n input:  "${result}"\n expect: "${expect}"\n  for highlighting "${input}" on "${text}"`);
        } else {
            console.log(`success: "${result}" has highlighted "${input}" on "${text}" properly.`);
        }
    }

    test(`Scott`, `c`, `S<b>c</b>ott`);
    test(`<<img src="xss" onerror="alert('xss')">`, `<`, `<b>&lt;</b><b>&lt;</b>img src=&quot;xss&quot; onerror=&quot;alert(&#39;xss&#39;)&quot;&gt;`);
})();
```

**Comment by codefactor** [comment]
I don't know if I'm doing this right - but here is a proposal how you might fix it.

https://github.com/SAP/ui5-webcomponents/pull/4246

I did not add unit tests, though, since I wasn't sure how to do that.

**Comment by LidiyaGeorgieva** [comment]
Hello @SAP/ui5-webcomponents-topic-rl,

Please have a look at this issue. It is reproducible.

Best Regards,
Lidiya

**Comment by ilhan007** [comment]
There is already a fix under review #4246, I will add reviewers from team Rila

**Comment by codefactor** [comment]
@MapTo0 ,
There is still one issue with the PR in terms of the unit test failing, and as reminder we need this to be patched next week as an emergency. If necessary you might want to purposely disable the unit test and maybe just start out on successfactors 0.31 branch and fix it on master later.
